manage-data/ingest/transform-enrich/readable-maintainable-ingest-pipelines.md

```diff
@@ -0,0 +1,677 @@
+
+Avoid chaining several `gsub` processors to strip units and manually convert values. This approach is error-prone, hard to maintain, and can easily miss edge cases.
+
+```json
+{
+  "gsub": {
+    "field": "document.size",
+    "pattern": "M",
+    "replacement": "",
+    "ignore_missing": true,
+    "if": "ctx?.document?.size != null && ctx.document.size.endsWith(\"M\")"
+  }
+},
+{
+  "gsub": {
+    "field": "document.size",
+    "pattern": "(\\d+)\\.(\\d+)G",
+    "replacement": "$1$200",
+    "ignore_missing": true,
+    "if": "ctx?.uws?.size != null && ctx.document.size.endsWith(\"G\")"
+  }
+},
+{
+  "gsub": {
+    "field": "document.size",
+    "pattern": "G",
+    "replacement": "000",
+    "ignore_missing": true,
+    "if": "ctx?.uws?.size != null && ctx.document.size.endsWith(\"G\")"
+  }
+}
+```
+
+### ![ ](../../images/icon-check.svg) **Do**: Use the `bytes` processor for automatic conversion
+
+The [`bytes` processor](https://www.elastic.co/guide/en/elasticsearch/reference/current/bytes-processor.html) automatically parses and converts strings like `"100M"` or `"2.5GB"` into their byte values. This is more reliable, easier to maintain, and supports a wide range of units.
+
+```json
+POST _ingest/pipeline/_simulate
+{
+  "docs": [
+    {
+      "_source": {
+        "document": {
+          "size": "100M"
+        }
+      }
+    }
+  ],
+  "pipeline": {
+    "processors": [
+      {
+        "bytes": {
+          "field": "document.size"
+        }
+      }
+    ]
+  }
+}
+```
+
+:::{tip}
+After storing values as bytes, you can use Kibana's field formatting to display them in a human-friendly format (KB, MB, GB, etc.) without changing the underlying data.
+:::
+
+## Rename fields
+
+The [rename processor](elasticsearch://reference/enrich-processor/rename-processor.md) renames a field. There are two flags:
+
+- `ignore_missing`: Useful when you are not sure that the field you want to rename exists.
+- `ignore_failure`: Helps with any failures encountered. For example, the rename processor can only rename to non-existing fields. If you already have the field `abc` and you want to rename `def` to `abc`, the operation will fail.
+
+## Script processor
+
+If no built-in processor can achieve your goal, you may need to use a [script processor](elasticsearch://reference/enrich-processor/script-processor.md) in your ingest pipeline. Be sure to write scripts that are clear, concise, and maintainable.
+
+### Add new fields
+
+All of the above discussed ways to [access fields](#access-fields) and retrieve their values is applicable within the script context. [Null handling](#null-safe-operators) is still an important aspect when accessing the fields.
+
+:::{tip}
+The fields API is the recommended way to add new fields.
+:::
+
+**Fields API**
+We get the following field `cpu.usage` and we want to rename it to `system.cpu.total.norm.pct` which represents a scale from 0-1.0, where 1 is the equivalent of 100%.
```

Comment:
Are we actually renaming a field or adding a new one? It looks like we're adding a new one, but let me know if I'm misinterpreting! 

```suggestion
For example, add a new `system.cpu.total.norm.pct` field based on the value of the `cpu.usage` field. The value of the existing `cpu.usage` field is a number on a scale of 0-100. The value of the new `system.cpu.total.norm.pct` field will be on a scale from 0-1.0 where 1 is the equivalent of 100 in the `cpu.usage` field.

**Option 1: Fields API (preferred)**
Create a new `system.cpu.total.norm.pct` field and set the value to the value of the `cpu.usage` field divided by `100.0`.
```